src/lib/tls/tls_algos.h

```diff
@@ -117,48 +118,81 @@ enum class Group_Params : uint16_t {
    HYBRID_SECP521R1_KYBER_1024_R3_OQS = 0x2F3D,
 };
 
-constexpr bool is_x25519(const Group_Params group) {
-   return group == Group_Params::X25519;
-}
+class Group_Params final {
+   public:
+      using enum Group_Params_Code;
 
-constexpr bool is_ecdh(const Group_Params group) {
-   return group == Group_Params::SECP256R1 || group == Group_Params::SECP384R1 || group == Group_Params::SECP521R1 ||
-          group == Group_Params::BRAINPOOL256R1 || group == Group_Params::BRAINPOOL384R1 ||
-          group == Group_Params::BRAINPOOL512R1;
-}
+      constexpr Group_Params() : m_code(Group_Params_Code::NONE) {}
 
-constexpr bool is_dh(const Group_Params group) {
-   return group == Group_Params::FFDHE_2048 || group == Group_Params::FFDHE_3072 || group == Group_Params::FFDHE_4096 ||
-          group == Group_Params::FFDHE_6144 || group == Group_Params::FFDHE_8192;
-}
+      constexpr Group_Params(Group_Params_Code code) : m_code(code) {}
 
-constexpr bool is_pure_kyber(const Group_Params group) {
-   return group == Group_Params::KYBER_512_R3 || group == Group_Params::KYBER_768_R3 ||
-          group == Group_Params::KYBER_1024_R3;
-}
+      constexpr Group_Params(uint16_t code) : m_code(static_cast<Group_Params_Code>(code)) {}
 
-constexpr bool is_hybrid(const Group_Params group) {
-   return group == Group_Params::HYBRID_X25519_KYBER_512_R3_CLOUDFLARE ||
-          group == Group_Params::HYBRID_X25519_KYBER_768_R3_CLOUDFLARE ||
-          group == Group_Params::HYBRID_X25519_KYBER_512_R3_OQS ||
-          group == Group_Params::HYBRID_X25519_KYBER_768_R3_OQS ||
-          group == Group_Params::HYBRID_SECP256R1_KYBER_512_R3_OQS ||
-          group == Group_Params::HYBRID_SECP384R1_KYBER_768_R3_OQS ||
-          group == Group_Params::HYBRID_SECP521R1_KYBER_1024_R3_OQS;
-}
+      /**
+      * Returns std::nullopt if an unknown name
+      */
+      static std::optional<Group_Params> from_string(std::string_view group_name);
 
-constexpr bool is_kem(const Group_Params group) {
-   return is_pure_kyber(group) || is_hybrid(group);
-}
+      constexpr bool operator==(Group_Params_Code code) const { return m_code == code; }
 
-constexpr bool is_post_quantum(const Group_Params group) {
-   return is_pure_kyber(group) || is_hybrid(group);
-}
+      constexpr bool operator==(Group_Params other) const { return m_code == other.m_code; }
+
+      constexpr bool operator<(Group_Params other) const { return m_code < other.m_code; }
```

Comment:
C++20 doesn't need to get this spelled out anymore. Simply default the spaceship operator:

```C++
constexpr auto operator<=>(const Group_Params&) const = default;
```

Godbolt: https://godbolt.org/z/4jovGh4EY

Reply:
:+1:

Reply:
Unfortunately that doesn't work here as it seems the additional overload of operator== (to handle `Group_Params_Code`) prevents the compiler from synthesizing the operator== we want using <=>